### PR TITLE
Group Equivariant Convolutional Networks

### PR DESCRIPTION
**Pull request by chrisrothUT**
Hi everybody,

I now have a working implementation of G-CNNs that works with a generic SymmGroup. GCNN works exactly like RBM_Symm. Here is a 4 layer GCNN on the Kagome lattice:

```
graph = nk.graph.Lattice(basis_vectors=[[1.,0.],[-1./2.,np.sqrt(3)/2]],extent=[2,2],atoms_coord=[[0.,0.],[-1./4.,np.sqrt(3)/4.],[1./2.,0.]])

ma = nk.models.GCNN(
    symmetries = graph,
    layers = 4,
    features = 4,
    dtype=complex
)
```

And we're off! The N layer model involves a layer DenseSymm, followed by N-1 DenseEquivariant layers which convolve over the entire set of poses based on their relative symmetry orientations. Like RBMSymm, this will produce identical outputs for each automorphism of the lattice, and I've included a similar test.

Questions
- What would be a good test for the layer DenseEquivariant?
- Can we figure out a way to only call reshape() once in DenseEquivariant (instead of twice)
- What are good defaults for initialization? 

Future
- Implement restricted convolutions. This requires information about the symmetry group, not just a randomly ordered list of automorphisms. 
- After testing on a variety of lattices, we tend to get caught in local minima on frustrated lattices. So we should think about adding an annealing method (i.e. optimizing phases first) to VMC (obviously beyond the scope of this PR)

### CODE REVIEW

File: netket/graph/lattice.py
Comment (PhilipVinc) [inline netket/graph/lattice.py:23]:
```suggestion
cutoff_tol = 1e-5
"""Tolerance for the maximum distance cutoff when computing the sparse distance matrix.
This is necessary because of floating-point errors when computing the distance in non-trivial 
lattices.
""" 

```

Comment (PhilipVinc) [reply netket/graph/lattice.py:23]: Add this here and use it below as needed.
This way it is more clear why we have it and we always use the same number (easier to change in the future if necessary).

File: netket/nn/linear.py
Comment (chrisrothUT) [reply netket/nn/linear.py:393]: I was under the impression that reshapes copy memory. At least that is the case for Pytorch where .view() was the analog of reshape that doesn't touch memory. I trust your read on the situation.

I can't think of a good way to avoid the transposition. The mapping is transformed_kernel[a*n_symm + i, b*n_symm + j] = kernel[group_algebra[i,j], a, b]

Comment (chrisrothUT) [reply netket/nn/linear.py:393]: right now group_algebra is a flattened and turned into a tuple

Comment (PhilipVinc) [reply netket/nn/linear.py:393]: I can't speak for torch as I never used it, but that is the case* in jax. (*: Technically it's not guaranteed as it's treated like an optimization, but the Jax compiler is very good at optimizing away memory reuses)

File: netket/nn/linear.py
Comment (chrisrothUT) [reply netket/nn/linear.py:443]: Yeah, it definitely feels a little clunky. The big issue IMO is that the Flax module only wants static attributes, probably something to do with the jit compilation. They're may be a workaround I don't know about

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py]: fixed

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:44]: should I add annotation in both places?

File: netket/graph/lattice.py
Comment (chrisrothUT) [reply netket/graph/lattice.py:29]: Yeah this makes sense

File: netket/graph/symmetry.py
Comment (PhilipVinc) [reply netket/graph/symmetry.py]: Yes.
They make the code clearer to whoever will read it in the future.

File: netket/nn/linear.py
Comment (PhilipVinc) [reply netket/nn/linear.py:443]: See what we settled with in #621.
You should get rid of `create_DenseEquivariant` and users should pass in a `SymmetryGroup` object or an `HashableArray` if they want to do something custom.

Comment (chrisrothUT) [reply netket/nn/linear.py:443]: Looks good. There's still a bit of a problem on my end. group_algebra() is slow and if I call it in setup.py, it will call it at every iteration

File: netket/graph/symmetry.py
Comment (PhilipVinc) [inline netket/graph/symmetry.py:124]: Why is this converted to a tuple? Is this in order to work with Flax?

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:64]: See the recent PR by @femtobit and changes to DenseSymm. Here you can just call 'symm.shape'

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:122]: We opted for not taking in a graph when symmetries are required. It's either a matrix or symmgroup. So remove this if branch

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:124]: Yeah, I've fixed it and made it a HashableArray

Comment (PhilipVinc) [reply netket/graph/symmetry.py:124]: Great. I just would like @femtobit opinion here about whoever it makes sense to return an hashable array here or if we should keep the build_GCN method and return a normal array here...

I think that if we make HashableArray behave satisfy the numpy API we might get away with just returning a HashableArray. 

This means adding dtype,shape,__len__,ndim,getitem I think...

Then you could use it as a numpy array and all is good as long as you don't modify it.

File: netket/models/equivariant.py
Comment (chrisrothUT) [reply netket/models/equivariant.py:122]: Yes, I read the PR and I've implemented the changes.  Now the constructors are gone and GCNN excepts a SymmGroup or a HashableArray. If it's an array, you also need to specify group_algebra

Comment (PhilipVinc) [reply netket/models/equivariant.py:122]: Yeah sorry, all is fine.
I guess i did not refresh the page... my bad

Comment (chrisrothUT) [reply netket/models/equivariant.py:122]: No problem, just didn't want to waste your time looking through my code when I was fixing things already :)

File: netket/graph/symmetry.py
Comment (femtobit) [reply netket/graph/symmetry.py:124]: > I think that if we make HashableArray behave satisfy the numpy API we might get away with just returning a HashableArray.

I wanted to avoid that. Making it behave like a numpy array _most of the time_ is doable, but there might be additional bugs or at least inconsistencies when it doesn't. (Just as `DeviceArray` and `ndarray` mostly behave the same but have subtle differences.)

The only time you need a `HashableArray` is when you use it as a model attribute. I feel we should just wrap it there and use normal objects otherwise.

Comment (femtobit) [reply netket/graph/symmetry.py:124]: Maybe it also makes sense for group algebra to be its own class? I haven't had time to look at this PR in more detail so I don't yet know if that makes any sense though. I'll take a closer look in the next days.

File: netket/models/equivariant.py
Comment (femtobit) [reply netket/models/equivariant.py:64]: Should be `np.asarray(self.symmetries).shape` as of the current master.

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:124]: I moved group_algebra to be a function in SymmGroup. I don't see why you would need to call it on something that isn't a SymmGroup object

**Comment by chrisrothUT** [comment]
Okay, made a few changes. GCNN takes symmetries=SymmGroup or arrays need to be specified for symmetries and group_algebra. I added a test for GCNN_creation as well

File: Test/Models/test_nn.py
Comment (femtobit) [inline Test/Models/test_nn.py:30]: You could just pass a function, like
```suggestion
def _setup_symm(symmetries, N, lattice=nk.graph.Chain):
     g = lattice(N)
```
and change the calling sites accordingly. That saves the additional `if` blocks.

File: Test/Models/test_rbm.py
Comment (femtobit) [inline Test/Models/test_rbm.py:59]:
```suggestion
@pytest.mark.parametrize("lattice", [nk.graph.Chain, nk.graph.Square])
```
see above

File: Test/Models/test_nn.py
Comment (chrisrothUT) [reply Test/Models/test_nn.py:30]: this works for me

**Comment by PhilipVinc** [comment]
You should also add the new classes/model to the documentation by putting their name in the correct subsection of docs/docs/api.rst

File: netket/nn/linear.py
Comment (chrisrothUT) [reply netket/nn/linear.py:364]: I tested tensordot with and without transpose, as well as jax.lax.dot and they all take about 3 microseconds per iteration on a Colab GPU. The intra run variance for N=1000 was larger than the individual differences between runs  

```
n_sym = 100
n_hid = 16
batch_size = 100

def transpose(inp,kernel):
    kernel = kernel.transpose(2,0,3,1)
    kernel = kernel.reshape(n_hid*n_sym,n_hid*n_sym)

    out = jax.lax.dot_general(inp,kernel,(((inp.ndim - 1,), (0,)), ((), ())))

    return out

def lax_dot(inp,kernel):
    kernel = kernel.transpose(2,0,3,1)
    kernel = kernel.reshape(n_hid*n_sym,n_hid*n_sym)

    out = jax.lax.dot(inp,kernel)

    return out

def no_transpose(inp,kernel):
    inp = inp.reshape(batch_size,n_hid,n_sym) 

    out = jax.lax.dot_general(inp,kernel,(((1,2), (2,0)), ((), ())))
    out = out.reshape(batch_size,n_sym*n_hid)

    return out

jit_transpose = jax.jit(partial(transpose,kernel=kernel_unshaped))
jit_no_transpose = jax.jit(partial(no_transpose,kernel=kernel_unshaped))
jit_dot = jax.jit(partial(lax_dot,kernel=kernel_unshaped))
```

Comment (PhilipVinc) [reply netket/nn/linear.py:364]: I think the XLA compiler is compiling the same code for both. So that is good.

File: netket/graph/symmetry.py
Comment (PhilipVinc) [inline netket/graph/symmetry.py:85]:
```suggestion
```

File: netket/graph/symmetry.py
Comment (PhilipVinc) [inline netket/graph/symmetry.py:105]:
```suggestion
```

**Comment by chrisrothUT** [comment]
Okay, I added a test for DenseEquivariant, and updated the docs. I think this should be ready for review

**Comment by chrisrothUT** [comment]
Okay there's still this annoying issue. SymmGroup.group_algebra() is an expensive function, where compute time scales like n_symm^3. If you pass SymmGroup to GCNN it computes group_algebra() in setup, and for whatever reason calls the function several times while its compiling. There's a few options for remedying this issue:

1. Speed up group_algebra(). Right now I use loops and there's probably some ways to take advantage of linear search algorithms
2. Give the user a warning. "This symmetry group has a lot of elements, for quicker compile time pass group_algebra as an argument directly" 
3. Call group_algebra in the init method of SymmGroup, but this might waste time for folks who don't want access to this tensor

**Comment by PhilipVinc** [comment]
how expensive is that function? I thought it was fairly cheap...
how long does it take on reasonably sized lattices?

In any case, as it's simply three nested loop, I'd slap a `@numba.njit` decorator on it to make us happier.
Also, since the SymmGroup is non mutable/static, you can just compile it once and then serve it cached, like we do for the hash

```python
@numba.njit
def group_algebra(automorphisms, inverse):
    return whatever

class SymmGroup:
   def __init__():
        ....
        self.__group_algebra = None

    def group_algebra(self):
         if self.__group_algebra is None:
            self.__group_algebra = HashableArray(group_algebra(self.automorphisms, self.inverse())
         return self.__group_algebra
```

You could also do the same for `inverse` and `to_array()` if you think they take considerable time.

**Comment by chrisrothUT** [comment]
For a 6x6 honeycomb with n_symm = 432 it takes 5 mins. So for a 10x10 lattice it probably takes a couple of hours. If we can find a way to make it take n_symm^2 time it shouldn't be a problem at all. 

This is a nice solution to the problem of calling it multiple times.

**Comment by chrisrothUT** [comment]
Ohh shoot, that can't work because SymmGroup is a frozen dataclass

**Comment by PhilipVinc** [comment]
Its legit, just use __setattr__

**Comment by gcarleo** [comment]
sorry I am a bit lost in the conversation, maybe can you tell me where ```SymmGroup.group_algebra()```is implemented and why it takes so long on small lattices? (if finding symmetries for a 10x10 lattice takes a few hours this is a very serious bottleneck we should solve, most likely going back to the automorphisms or finding a faster implementation!)

File: netket/graph/symmetry.py
Comment (gcarleo) [inline netket/graph/symmetry.py:117]: oh I see, it's this triple loop + an if running an all the vector!!... uhm isn't there a way to rewrite this is a faster way?

**Comment by PhilipVinc** [comment]
Also @chrisrothUT , while this PR is in a good state (assuming we can address the N^3 scaling), would you consider also adding a commented tutorial (jupyter notebook) walking people through the model and the parameters, how it works and what it does, and maybe comparing it to other networks? A bit like https://www.netket.org/tutorials/Heisenberg1d.html

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:117]: I'm sure there's a solution with hashing that is a lot faster. I'm trying to identify the indices of duplicates between a list of length of N and a list of length N^2 (which I can make with arrays). I think I can make a dict with the N^2 list and loop over the N list?

**Comment by chrisrothUT** [comment]
@PhilipVinc I already wrote something for my blog so I can make that into a tutorial!

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:117]: Yeah, I think I have to hash over the length N list but this should still be fast enough

Comment (chrisrothUT) [reply netket/graph/symmetry.py:117]: Okay, problem has been solved, it's a lot faster than graph.automorphisms(), and scales approximately like 17microseconds*n^2. So even for a 10x10 triangular that is like 25 seconds

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:36]: Can you also write something more here?

Put a new line and then write a bit more, so the website documentation is actually helpful.
you can also write latex equations.

File: netket/graph/symmetry.py
Comment (PhilipVinc) [inline netket/graph/symmetry.py:88]: can you explain this better?

File: netket/graph/symmetry.py
Comment (PhilipVinc) [reply netket/graph/symmetry.py:117]: Still, could you check if putting it in a numba jit block makes it even a bit faster?

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:48]:
```suggestion
    group_algebra: Optional[HashableArray] = None
```

File: netket/models/equivariant.py
Comment (chrisrothUT) [reply netket/models/equivariant.py:36]: Okay that is good to know I'll try to make these more helpful

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:50]:
```suggestion
    Does not need to be specified if symmetries is a SymmGroup
```

**Comment by chrisrothUT** [comment]
@PhilipVinc Considering group algebra is significantly faster than automorphisms(), do you still think it's worth it to make the changes we discussed

File: netket/nn/linear.py
Comment (PhilipVinc) [inline netket/nn/linear.py:323]: maybe i am ignorant, but what is a pose of a symmetry group?

File: netket/nn/linear.py
Comment (PhilipVinc) [inline netket/nn/linear.py:324]: identical_shape but `[n_symm*in_features]` and `[n_symm*out_features]` are clearly different?
What do you mean?

File: netket/nn/linear.py
Comment (PhilipVinc) [inline netket/nn/linear.py:329]: typo?

File: netket/nn/linear.py
Comment (PhilipVinc) [inline netket/nn/linear.py:325]: Maybe you could also give an explicit example of what happens to shapes. 
Given an input shape (N_batches, N_spins, blablabla) you get ...

File: netket/nn/linear.py
Comment (PhilipVinc) [inline netket/nn/linear.py:332]: I'm not sure that this is correct. The input/output size will also depend on batches... I think giving an explicit example with what the shapes should be would be clearer.

File: netket/nn/linear.py
Comment (chrisrothUT) [reply netket/nn/linear.py:323]: Pose is an ML term, is the word orientation more clear? It's quite tricky to explain what is going on in a paragraph, that's why I plan to write more stuff

File: netket/nn/linear.py
Comment (chrisrothUT) [reply netket/nn/linear.py:324]: Good catch

**Comment by PhilipVinc** [comment]
If its really not the bottleneck, ok. Let's add it to the todo and try to speed  up automorphisms @femtobit with numba in the future.

File: netket/models/equivariant.py
Comment (PhilipVinc) [reply netket/models/equivariant.py:36]: A good example of a complete documentation and using latex is in LocalLiouvillian.

In general everything you write before a newline is considered the short-summary used in short descriptions in the website, and then after the newline you can write everything else, generates the docpage.

File: netket/nn/linear.py
Comment (PhilipVinc) [reply netket/nn/linear.py:323]: I can only guess what you mean by orientation...

As a short description I would specify carefully what should be the inputs and outputs and with maybe easy/to/understand words what it means. 
Then newline, 
then try to explain concepts you need to say what this does. 
Links to websites/papers are welcome 
you write a link with the syntax 
```python
` text <https:://link.to.something>`_
```

**Comment by chrisrothUT** [comment]
Okay I think I've covered everything, I'm taking the afternoon off but I will do the tutorial tomorrow!

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
This looks good to me.
I'll wait for a review by @femtobit too.

The tutorial (to be added in tutorials/ folder and also in the list of tutorials in the `docs/turoails.rst`) would be great, but I think we can merge it later when it's ready.

**Comment by femtobit** [comment]
Yes, sorry about letting you wait but I'd like to look at this PR and the group algebra a bit more closely and didn't get around to it yet. I hope I can take a look soon (definitely in the next days).

**Comment by chrisrothUT** [comment]
No problem at all, take your time. I plan to finish the tutorial today, so maybe we can add that in before the merge

File: netket/graph/lattice.py
Comment (femtobit) [inline netket/graph/lattice.py:31]: This innocent-looking change changes the time it takes for `Test/Graph/test_graph.py::test_automorphisms` by _a lot_:
```
# before:
2.13s call     Test/Graph/test_graph.py::test_automorphisms
# after:
226.23s call     Test/Graph/test_graph.py::test_automorphisms
```

This is probably because the lattice graphs previously did not recognize all edges. In any case, we should probably fix this test.

By the way, is there a reason `test_automorphism` is only run on the lattices? Also, do we really need this test in the current form? At the moment, it only seems to check that two different `networkx` methods return the same results (as all our graphs use networkx anyway for this).

Comment (PhilipVinc) [reply netket/graph/lattice.py:31]: I'm all in for writing better tests.
I don't know who wrote them originally.

File: netket/graph/symmetry.py
Comment (femtobit) [inline netket/graph/symmetry.py:104]: I'm not sure "group algebra" is a clear name for this function. My group theory knowledge is somewhat rusty, but (aside from using `g^-1`) doesn't it essentially compute a representation of the group's Cayley table (i.e., it's multiplication table)?

Or is there an established meaning of "group algebra" that coincides with what this function does? Otherwise I think we should try and find a different name.

File: netket/graph/symmetry.py
Comment (femtobit) [inline netket/graph/symmetry.py:88]: Maybe like this?
```suggestion
        the original symmetry element. If :code:`g = self[element]` and :code:`h = self.inverse()[element]`,
        then :code:`gh = product(g, h)` will act as the identity on the sites of the graph, i.e., :code:`np.all(gh(sites) == sites)`.
```

File: netket/graph/lattice.py
Comment (PhilipVinc) [reply netket/graph/lattice.py:31]: There is overlap with what is being discussed [here](https://github.com/netket/netket/pull/633#discussion_r616672597)

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:104]: It's a Cayley table but one side is the involution of the group. I'm open to renaming it though.

File: netket/graph/lattice.py
Comment (femtobit) [reply netket/graph/lattice.py:31]: I think for now we could honestly just remove that test. At least to some extent the `NetworkX.autmorphisms` method at least is covered in the symmetry group tests as well, so we would not lose significant coverage.

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:104]: How about flattenen_Cayley(). And something like returns a flattened Cayley table, s.t. for u = flattenen_Cayley()[elem1*n_symm + elem2],  g = self.inverse()[elem1],  and h = self[elem2], u = gh

Comment (chrisrothUT) [reply netket/graph/symmetry.py:104]: actually it would be u = self[flattenen_Cayley()[elem1*n_symm + elem2]], maybe this is too cumbersome

Comment (chrisrothUT) [reply netket/graph/symmetry.py:104]: *flattened

Comment (femtobit) [reply netket/graph/symmetry.py:104]: Now that you mention it: Is there a reason the flattened table is returned? It sounds more generally useful for me to return the `ndim=2` table. Flattening it when needed is easier than the reverse operation.

File: netket/graph/symmetry.py
Comment (femtobit) [inline netket/graph/symmetry.py:138]: I think this should return an array. The only place I'd use `HashableArray` is as an attribute of a frozen dataclass (which includes our models). Otherwise it's not worth the hassle.

File: netket/graph/symmetry.py
Comment (femtobit) [reply netket/graph/symmetry.py]: Now that you mention it: Is there a reason the flattened table is returned? It sounds more generally useful for me to return the `ndim=2` table. Flattening it when needed is easier than the reverse operation.

Comment (chrisrothUT) [reply netket/graph/symmetry.py]: Okay I've gone ahead and changed the name to flattened_Cayley, and rewrote the description with that in mind

Comment (femtobit) [reply netket/graph/symmetry.py]: Cayley table is I think a bit too much group theory jargon (especially given that it's not exactly the table of `g*h` due to the inverse) for the name. How about just `product_table`?

File: netket/graph/symmetry.py
Comment (femtobit) [reply netket/graph/symmetry.py:89]: I find it sufficiently clear now.

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py]: product table sounds fine. I guess I could just return an array of n_dim = 2 and flatten/make it Hashable inside of setup()?

File: netket/graph/symmetry.py
Comment (femtobit) [inline netket/graph/symmetry.py:102]: I wonder which is better: Returning a `SymmGroup` here or just the indices of the inverse elements (i.e., `inverse` directly). The indices could be a bit more generally useful and are harder to reconstruct from `.elems` later.

Comment (chrisrothUT) [reply netket/graph/symmetry.py:102]: I've vacilated on this. I'm happy to return the indices

Comment (femtobit) [reply netket/graph/symmetry.py:102]: Yes, let's return the indices.

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py:137]: Okay I'll fix this

File: netket/graph/symmetry.py
Comment (chrisrothUT) [reply netket/graph/symmetry.py]: Actually I'm just going to have DenseEquivariant take flattened_product_table and complain if it gets something else. GCNN will do the same if product_table is specified, and if it just gets SymmGroup

Comment (femtobit) [reply netket/graph/symmetry.py]: > product table sounds fine. I guess I could just return an array of n_dim = 2 and flatten/make it Hashable inside of setup()?

How about this: The `DenseEquivariant` layer accepts `Union[HashableArray, SymmGroup]` and in setup you either use the array directly or, if the attribute is a `SymmGroup` call something like `group.product_table().ravel()` and work with that?

Then everything is consistent with `DenseSymm`.

Comment (chrisrothUT) [reply netket/graph/symmetry.py]: Okay great, I will do that

File: netket/graph/lattice.py
Comment (PhilipVinc) [reply netket/graph/lattice.py:31]: can you open a PR for this?

**Review by femtobit — APPROVED** [review]
This looks pretty good to me now, thanks @chrisrothUT. As far as I am concerned this can be merged.

An example would be nice but since the plan is to add a tutorial, that will be addressed then anyways.

File: netket/graph/symmetry.py
Comment (PhilipVinc) [reply netket/graph/symmetry.py]: It looks to me this was not reverted to indices?

Comment (PhilipVinc) [reply netket/graph/symmetry.py]: nevermind, i was on an old diff

File: Test/Models/test_nn.py
Comment (attila-i-szabo) [inline Test/Models/test_nn.py:95]:
```suggestion
    # output should be equivariant
    assert jnp.allclose(dot(out, sym_op), out_trans)
```
see https://github.com/netket/netket/issues/698#issuecomment-834948357 for details

**Pull request merged**
